Activate streaming examples in CI

franzpoeschel opened the pull request:
So far, the streaming examples are commented out because our CI couldn't deal with that. Writers and Readers need to be executed asynchronously. If executing them sequentially, like our CI does, one of them will wait for the other to show up indefinitely.

Activate them.

TODO
- [x] Fix things in the CI so things don't hang.
- [x] CI runs fine on Windows, but ADIOS2 SST engine doesn't. Changing extension to `bp` makes the examples work.
    EDIT: SST [unsupported on MSVC](https://github.com/ornladios/ADIOS2/blob/5948ca8a85e05eaf2ff07c6b64fa049fe0c4f9bb/cmake/DetectOptions.cmake#L320). So, deactivate Streaming examples on Windows.

franzpoeschel committed:
Activate streaming examples

franzpoeschel committed:
Deactivate streaming examples in CI

franzpoeschel committed:
Run streaming examples asynchronously

Probably working on Linux only

franzpoeschel committed:
Make things more portable

franzpoeschel committed:
Deactivate streaming examples on Windows

ax3l commented:
@franzpoeschel can you please add, after a double white-space at the end-of-the-line, a `# lgtm [py/unreachable-statement]` to silence the Linux LGTM warnings?

ax3l commented on examples/10_streaming_write.cpp:
```diff
@@ -1,4 +1,9 @@
-#define BUILD_STREAMING_EXAMPLE false
+#ifdef _WIN32
```
Is this a general windows issue or a problem with MSVC?
(Note: we also ship and often prefer Clang on Windows)

franzpoeschel replied:
[Just with MSVC](https://github.com/ornladios/ADIOS2/blob/5948ca8a85e05eaf2ff07c6b64fa049fe0c4f9bb/cmake/DetectOptions.cmake#L320). Should I just drop the check and do a better check in the ADIOS2 backend?

ax3l replied:
Yes, that would be better I think.
Let's also add a link to that place as a comment.

franzpoeschel replied:
done

ax3l commented on CMakeLists.txt:
```diff
@@ -1050,6 +1052,11 @@ if(openPMD_BUILD_TESTING)
                 endforeach()
             endif()
         endif()
+        if(openPMD_HAVE_ADIOS2)
```
If the issue is only in execution, you could skip the `add_test` here for Windows or try to search for `sh` and skip it conditionally if the command is not found:
https://github.com/ECP-WarpX/WarpX/blob/21.01/cmake/WarpXFunctions.cmake#L1-L12

franzpoeschel replied:
`sh` on Windows is no issue actually. I didn't expect it, but it seems to work.

franzpoeschel committed:
Allow to ask backends for properties

franzpoeschel committed:
Only run SST examples if SST is available

franzpoeschel committed:
transform to lowercase in Series class already

franzpoeschel committed:
Add link that explains why no SST in MSVC

https://github.com/ornladios/ADIOS2/blob/5948ca8a85e05eaf2ff07c6b64fa049fe0c4f9bb/cmake/DetectOptions.cmake#L320

franzpoeschel committed:
Add missing import to Python

franzpoeschel committed:
Two little syntax errors

ax3l commented on examples/10_streaming_write.cpp:
```diff
@@ -49,7 +52,4 @@ main()
-#else
-int main(){ return 0; }
-#endif
+}
```
```suggestion
}

```

ax3l commented on src/binding/python/Series.cpp:
```diff
@@ -169,6 +169,7 @@ void init_Series(py::module &m) {
         .def("flush", &Series::flush)
 
         .def_property_readonly("backend", &Series::backend)
+        .def("backend_property", &Series::backendProperty)
```
Just wondering if we need a new `series` property. We already have `openPMD::getFileExtensions()`/`openPMD_api.file_extensions` where we still need to add `.sst`, I think.

Do you have concrete additional plans to use this new property for?

franzpoeschel replied:
Didn't think of `getFileExtensions()`, I can use that one too.

ax3l commented on src/IO/ADIOS/ADIOS2IOHandler.cpp:
```diff
@@ -2042,8 +2042,25 @@ ADIOS2IOHandler::ADIOS2IOHandler(
+{
+    if( property == "sst" )
+    {
+#if defined( __GNUG__ ) || defined( __clang__ )
```
This looks a bit empiric - is there an ADIOS2 property/define that we could query for SST capability?

franzpoeschel replied:
I have found no such thing unfortunately. Can ask this tomorrow in the VC though.

ax3l replied:
Sounds good! Yes, one challenge is also that SST support in ADIOS2 depends on the binary variant we build for ADIOS2. So you can get a fully valid ADIOS2 install in some cases that has SST disabled. We might need a runtime (or at least compile-time) query for that from ADIOS2.

franzpoeschel committed:
Revert "Allow to ask backends for properties"

This reverts commit fa4a965e962024de20bd7df58466c8eaf9256dad.

franzpoeschel committed:
Remove unused imports

franzpoeschel committed:
Don't test SST in serial tests

franzpoeschel committed:
Fix SST condition

franzpoeschel committed:
Use ADIOS2_HAVE_SST macro

franzpoeschel committed:
Open Series only after checking we have SST

ax3l reviewed (approved):
Looking good, thank you! :sparkles:
